plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java

```diff
@@ -0,0 +1,330 @@
+            return object.executeMediaAsInputStream();
+        }
+
+        @Override
+        public void writeBlob(String bucketName, String blobName, InputStream inputStream, long blobSize) throws IOException {
+            logger.debug("writing blob [{}] in bucket [{}]", blobName, bucketName);
+            InputStreamContent stream = new InputStreamContent(null, inputStream);
+            stream.setLength(blobSize);
+
+            Storage.Objects.Insert insert = client.objects().insert(bucketName, null, stream);
+            insert.setName(blobName);
+            insert.execute();
+        }
+
+        @Override
+        public void deleteBlob(String bucketName, String blobName) throws IOException {
+            logger.debug("deleting blob [{}] in bucket [{}]", blobName, bucketName);
+            client.objects().delete(bucketName,blobName).execute();
+        }
+
+        @Override
+        public void deleteBlobs(String bucketName, Collection<String> blobNames) throws IOException {
+            if (blobNames == null || blobNames.isEmpty()) {
+                return;
+            }
+
+            logger.debug("deleting [{}] blob(s) in bucket [{}]", blobNames.size(), bucketName);
+            if (blobNames.size() == 1) {
+                deleteBlob(bucketName, blobNames.iterator().next());
+                return;
+            }
+
+            final Iterator<String> it = blobNames.iterator();
+            final Set<String> errors = ConcurrentCollections.newConcurrentSet();
+            BatchRequest batch = null;
+            int count = 0;
+
+            while (it.hasNext()) {
+                String blobName = it.next();
+                try {
+                    if (batch == null) {
+                        batch = client.batch();
+                    }
+                    logger.trace("add blob [{}] to batch request queue", blobName);
+                    client.objects().delete(bucketName, blobName).queue(batch, new JsonBatchCallback<Void>() {
+                        @Override
+                        public void onFailure(GoogleJsonError e, HttpHeaders responseHeaders) throws IOException {
+                            logger.trace("error when deleting blob [{}] in bucket [{}]", blobName, bucketName);
+                            errors.add("error [code: " + e.getCode() +", message: " + e.getMessage());
+                        }
+
+                        @Override
+                        public void onSuccess(Void aVoid, HttpHeaders responseHeaders) throws IOException {
+                            logger.trace("successfully deleted blob [{}] in bucket [{}]", blobName, bucketName);
+                        }
+                    });
+
+                    count += 1;
+                } catch (IOException e) {
+                    logger.error("Unable to queue delete request for blob [{}] in bucket [{}]", e, blobName, bucketName);
+                } finally {
+                    it.remove();
```

Comment:
That seems brittle (it assumes that collection we got is mutable) and not very useful (we don't rely on the fact that we deleted processed blobs from the collection).